`stylus-gas` edit (#562)
OffchainLabs/arbitrum-docs: arbitrum-docs/stylus/concepts/stylus-gas.md

---

> title: 'Gas and ink (Stylus)'

Author: intent: drive search traffic about "gas" and "ink" to this doc; drive traffic about "costs" and "pricing" to the reference doc

> title: 'Gas and ink (Stylus)'

Reviewer: ```suggestion
title: 'Gas and ink in Stylus'
```

> sidebar_label: 'Gas and ink'

Reviewer: ```suggestion
sidebar_label: 'Gas and ink in Stylus'
```
Kind of redundant, but since this doc is in a folder with something non-Stylus related, I think it's worth making explicitly clear this is Stylus related

Author: Everything in this folder is Stylus-specific, so I'd probably lean into the "keep "(Stylus)" qualifiers out of the sidebar" norm - I have faith that our readers will notice where these docs are in the sidebar ^^

description: 'A conceptual overview of gas and ink, the primitives that Stylus uses to measure the cost of WASM activation, compute, memory, and storage.'
author: rachel-bousfield
sme: rachel-bousfield
target_audience: 'Developers deploying smart contracts using Stylus.'
sidebar_position: 3
---

> **Gas and ink** are the pricing primitives that Stylus uses to determine how much it costs to handle specific opcodes and host I/Os. For an overview of specific opcode and host I/O costs, see [Opcode and host I/O pricing](/stylus/reference/opcode-hostio-pricing).

Reviewer: ```suggestion
**Ink** is a new pricing primitive that Stylus uses to determine how much it costs to handle specific opcodes and host I/Os, which is needed as many WASM opcodes cost less than `1 gas`. For an overview of specific opcode and host I/O costs, see [Opcode and host I/O pricing](/stylus/reference/opcode-hostio-pricing).
```

Reviewer: Definitely open to continue chiseling this line

Author: I'd recommend keeping this as-is - the title of the document sets the expectation that this explains Gas and Ink, which it does.

We can distinguish between gas and ink, and elaborate on the rationale underlying the ink primitive, within the body of the page :)

import PublicPreviewBannerPartial from '../partials/_stylus-public-preview-banner-partial.md';

<PublicPreviewBannerPartial />

> <!-- todo as a future optimization: pull the "Ink and gas" content up top; give devs what they need to know in order to build stuff - the "just in time" information - progressively disclose the "just in case" details. An example flow:

Author: for future consideration

Reviewer: Like a tldr / in a nutshell? I like that

Author: More of a refactoring that _incrementally introduces_ precision and complexity - which is similar to the "in a nutshell" pattern! We could even explicitly incorporate that here.

There are, however, minor overheads to using Stylus that may matter to your application:

> - The first time a WASM is deployed, it must be _activated_. This is generally a few million gas, though to avoid testnet DoS, we've set it to a fixed 14 million. Note that you do not have to activate future copies of the same program. For example, the same NFT template can be deployed many times without paying this cost more than once. We will soon make the fees paid depend on the program, so that the gas used is based on the complexity of the WASM instead of this very conservative, worst-case estimate.

Author: result of `yarn format`, still italicized

### The ink price

The ink price, which measures the amount of ink a single EVM gas buys, is configurable by the chain owner. By default, the exchange rate is `1:10000`, but this may be adjusted as the EVM and Stylus VM improve over time.

For example, if the Stylus VM becomes 2x faster, instead of cutting the nominal cost of each operation 2x, the ink price may instead be doubled, allowing 1 EVM gas to buy twice as much ink. This provides an elegant mechanism for smoothly repricing resources between the two VMs as each makes independent progress.

### User experience

It is important to note that users never need to worry about this notion of ink. Receipts will always be measured in gas, with the exchange rate applied automatically under the hood as the VMs pass execution back and forth.

However, developers optimizing contracts may choose to measure performance in ink to pin down the exact cost of executing various routines. The <a href="https://docs.rs/stylus-sdk/0.3.0/stylus_sdk/evm/fn.ink_left.html"><code>ink_left</code></a> function exposes this value, and various methods throughout the Rust SDK optionally accept ink amounts too.

> ### See also

Author: "end of document" pattern
